src/hotspot/share/jfr/periodic/sampling/jfrCPUTimeThreadSampler.cpp

```diff
@@ -0,0 +1,889 @@
+  void set_timer_time(timer_t timerid, int64_t period_millis);
+  void on_javathread_terminate(JavaThread* thread);
+
+  void handle_timer_signal(void* context);
+  void init_timers();
+  void stop_timer();
+};
+
+
+JfrCPUTimeThreadSampler::JfrCPUTimeThreadSampler(double rate, bool autoadapt, u4 max_traces, u4 max_frames_per_trace) :
+  _sample(),
+  _sampler_thread(nullptr),
+  _queues(max_traces, max_frames_per_trace),
+  _rate(rate),
+  _autoadapt(autoadapt),
+  _current_sampling_period_ns(compute_sampling_period(rate)),
+  _max_frames_per_trace(max_frames_per_trace),
+  _disenrolled(true),
+  _jfrFrames(JfrCHeapObj::new_array<JfrStackFrame>(_max_frames_per_trace)) {
+  assert(rate >= 0, "invariant");
+}
+
+JfrCPUTimeThreadSampler::~JfrCPUTimeThreadSampler() {
+  JfrCHeapObj::free(_jfrFrames, sizeof(JfrStackFrame) * _max_frames_per_trace);
+}
+
+void JfrCPUTimeThreadSampler::on_javathread_create(JavaThread* thread) {
+  if (thread->is_Compiler_thread()) {
+    return;
+  }
+  if (thread->jfr_thread_local() != nullptr) {
+    timer_t timerid;
+    if (create_timer_for_thread(thread, timerid) && thread->jfr_thread_local() != nullptr) {
+      thread->jfr_thread_local()->set_timerid(timerid);
+    }
+  }
+}
+
+void JfrCPUTimeThreadSampler::on_javathread_terminate(JavaThread* thread) {
+  if (thread->jfr_thread_local() != nullptr && thread->jfr_thread_local()->has_timerid()) {
+    timer_delete(thread->jfr_thread_local()->timerid());
+    thread->jfr_thread_local()->unset_timerid();
+  }
+}
+
+void JfrCPUTimeThreadSampler::start_thread() {
+  if (os::create_thread(this, os::os_thread)) {
+    os::start_thread(this);
+  } else {
+    log_error(jfr)("Failed to create thread for thread sampling");
+  }
+}
+
+void JfrCPUTimeThreadSampler::enroll() {
+  if (Atomic::cmpxchg(&_disenrolled, true, false)) {
+    log_info(jfr)("Enrolling CPU thread sampler");
+    _sample.signal();
+    init_timers();
+    log_trace(jfr)("Enrolled CPU thread sampler");
+  }
+}
+
+void JfrCPUTimeThreadSampler::disenroll() {
+  if (Atomic::cmpxchg(&_disenrolled, false, true)) {
+    log_info(jfr)("Disenrolling CPU thread sampler");
+    stop_timer();
+    Atomic::store(&_stop_signals, true);
+    while (_active_signal_handlers > 0) {
+      // wait for all signal handlers to finish
+      os::naked_short_nanosleep(1000);
+    }
+    _sample.wait();
+    _queues.reset();
+    Atomic::store(&_stop_signals, false);
+    log_trace(jfr)("Disenrolled CPU thread sampler");
+  }
+}
+
+void JfrCPUTimeThreadSampler::run() {
+  assert(_sampler_thread == nullptr, "invariant");
+  _sampler_thread = this;
+  int64_t last_autoadapt_check = os::javaTimeMillis();
+  while (true) {
+    if (!_sample.trywait()) {
+      // disenrolled
+      _sample.wait();
+    }
+    _sample.signal();
+
+    if (os::javaTimeMillis() - last_autoadapt_check > AUTOADAPT_INTERVAL_MS) {
+      autoadapt_period_if_needed();
+      last_autoadapt_check = os::javaTimeMillis();
+    }
```

Comment:
Do you really want wall-clock time here ???

Reply:
What else should I use? I want to check for the adaption of the sampling period every few seconds.

Reply:
Perhaps `os::javaTimeNanos()` would be better? It provides global monotonic time.
OTOH, `os::javaTimeMillis()` can move unexpectedly - a good example is the daylight saving time where the millis would suddenly jump 1 hour to the past.

Reply:
Good point, I totally forgot about the monotonicity.

Reply:
DST changes are not reflected in the low-level clocks, they come from higher-level API's that format times and account for TZ.

The clock underpinning `javaTimeMillis` can still jump due to NTP adjustments and so a monotonic clock source is preferable.

Reply:
So this issue is now resolved?

Reply:
Yes using nanoTime resolves any concerns.